File: Algorithm/Heap.java
package com.wumengyangok;

/**
 * Created by wumengyang on 26/02/2016.
 */
public class Heap {

    private int[] heap;
    private int length;
    private int maximum;

    Heap(int maximumLength) {
        this.maximum = maximumLength;
        heap = new int[maximum];
        length = 0;
    }

    private void downAdjust(int index) {

        int temp = heap[index];
        while((index * 2 + 1 <= length) && ((temp < heap[index * 2]) || (temp < heap[index * 2 + 1]))) {
            if (heap[index * 2] > heap[index * 2 + 1]) {
                heap[index] = heap[index * 2];
                index = index * 2;
            } else {
                heap[index] = heap[index * 2 + 1];
                index = index * 2 + 1;
            }
        }
        if (index * 2 == length) {
            if (temp < heap[length]) {
                heap[index] = heap[length];
                heap[length] = temp;
            } else {
                heap[index] = temp;
            }
        } else {
            heap[index] = temp;
        }

    }

    public void buildHeap() {
        for (int i = length / 2; i >= 1; i--) {
            downAdjust(i);
        }
    }

    public void addToTail(int newOne) {
        length++;
        try {
            heap[length] = newOne;
        } catch (ArrayIndexOutOfBoundsException e) {
            System.out.println("Out of bound");
        }

    }

    public void addAndMaintainHeap(int newOne) {

        addToTail(newOne);
        int index = length;
        int temp = heap[index];
        while ((index / 2 > 0) && (heap[index / 2] < temp)) {
            heap[index] = heap[index / 2];
            index = index / 2;
        }
        heap[index] = temp;

    }

    public void sort() {

        int saveLength = length;
        buildHeap();
        for (int i = saveLength; i >= 1; i--) {
            int temp = heap[1];
            heap[1] = heap[length];
            heap[length] = temp;
            length--;
            downAdjust(1);
        }
        length = saveLength;

    }

    public void output() {
        for (int i = 1; i <= length; i++) {
            System.out.printf("%d ",heap[i]);
        }
        System.out.println();
    }

}

File: Algorithm/ImplementationPy/heap.py
class Heap(object):
    def __init__(self, heap):
        self.heap = heap
        self.length = len(heap)

    def __add__(self, other):
        list.append(self.heap, other)
        self.length += 1

    def __output__(self):
        print(self.heap)

    def __shift_down__(self, index):
        temp = self.heap[index]
        i = index
        j = 2 * i + 1
        n = self.length
        while j < n:
            if (j < n - 1) and (self.heap[j] < self.heap[j+1]):
                j += 1
            if temp < self.heap[j]:
                self.heap[i] = self.heap[j]
                i = j
                j = 2 * j + 1
            else:
                break
        self.heap[i] = temp

    def __build_heap__(self):
        i = self.length / 2
        i = int(i)
        while i >= 0:
            self.__shift_down__(i)
            i -= 1

    def __shift_up__(self):
        index = self.length - 1
        temp = self.heap[index]
        while (int(index / 2) >= 0) and (self.heap[int(index / 2)] < temp):
            self.heap[index] = self.heap[int((index - 1) / 2)]
            if index == 0:
                break
            index = int((index - 1) / 2)
        self.heap[index] = temp


aHeap = Heap([])
for i in range(1, 16):
    aHeap.__add__(i)
aHeap.__build_heap__()
aHeap.__output__()